servant-server/src/Servant/Server/Internal/RoutingApplication.hs

```diff
@@ -240,13 +248,19 @@ passToServer Delayed{..} x =
 runDelayed :: Delayed env a
            -> env
            -> Request
-           -> IO (RouteResult a)
-runDelayed Delayed{..} env = runDelayedIO $ do
-  c <- capturesD env
-  methodD
-  a <- authD
-  b <- bodyD
-  DelayedIO (\ req -> return $ serverD c a b req)
+           -> IO (RouteResult a, IO ())
+runDelayed Delayed{..} env req = do
+  cleanupRef <- newIORef (return ())
+  routeRes <- runDelayedIO
+    (do c <- capturesD env
+        methodD
+        a <- authD
+        b <- bodyD
+        liftIO (writeIORef cleanupRef $ cleanupD b)
```

Comment:
Note that this is going to change a bit. @jkarni pointed out that adding clean up actions can happen from within `bodyD` and that we should not wait for it to terminate in order to add a clean up action, because if an exception occurs the cleanup never gets run. Particularly annoying in the case of file uploads where files can pile up in /tmp.